himbaechel/uarch/ng-ultra/ng_ultra.cc

```diff
@@ -0,0 +1,1027 @@
+        if (cluster->constr_abs_z) {
+            // Coerce root to absolute z constraint
+            root_loc.z = cluster->constr_z;
+            root_bel = ctx->getBelByLocation(root_loc);
+            if (root_bel == BelId() || !this->isValidBelForCellType(root_cell->type, root_bel))
+                return false;
+        }
+        placement.emplace_back(root_cell, root_bel);
+        return getChildPlacement(cluster, root_loc, placement);
+    });
+}
+
+BoundingBox NgUltraImpl::getRouteBoundingBox(WireId src, WireId dst) const
+{
+    if ((tile_type(src.tile)!=TILE_EXTRA_FABRIC || tile_type(dst.tile)!=TILE_EXTRA_FABRIC)) {
+        return { 0, 0, ctx->getGridDimX(), ctx->getGridDimY() };
+    }
+    int x0, y0, x1, y1;
+    auto expand = [&](int x, int y) {
+        x0 = std::min(x0, x);
+        x1 = std::max(x1, x);
+        y0 = std::min(y0, y);
+        y1 = std::max(y1, y);
+    };
+    tile_xy(ctx->chip_info, src.tile, x0, y0);
+    x1 = x0;
+    y1 = y0;
+    int dx, dy;
+    tile_xy(ctx->chip_info, dst.tile, dx, dy);
+    expand(dx, dy);
+    // Reach 7 MESH above and below (3 left and 3 right of TILE/CGB)
+    return {(x0 & 0xfffc) - 3*4, // 3 MESH on left 
+            (y0 & 0xfffc) - 4,   // row above
+            (x1 & 0xfffc) + 4*4, // MESH bellow and 3 right
+            (y1 & 0xfffc) + 8};  // current and row bellow
+}
+
+void NgUltraImpl::expandBoundingBox(BoundingBox &bb) const
+{
+    // Updated numbers for NG-ULTRA only
+    // x0 and y0 substract 1 TILE
+    // x1 and y1 adds 1 TILE (and of one added)
+    bb.x0 = std::max((bb.x0 & 0xfffc) - 4, 0);
+    bb.y0 = std::max((bb.y0 & 0xfffc) - 4, 0);
+    bb.x1 = std::min((bb.x1 & 0xfffc) + 8, ctx->getGridDimX());
+    bb.y1 = std::min((bb.y1 & 0xfffc) + 8, ctx->getGridDimY());
+}
+
+delay_t NgUltraImpl::estimateDelay(WireId src, WireId dst) const
+{
+    int sx, sy, dx, dy;
+    tile_xy(ctx->chip_info, src.tile, sx, sy);
+    tile_xy(ctx->chip_info, dst.tile, dx, dy);
+    if (sx==dx && sy==dy) {
+        // Same sub tile
+        return 50;
+    } else if (((sx & 0xfffc) == (dx & 0xfffc)) && ((sy & 0xfffc) == (dy & 0xfffc))) {
+        // Same "TILE"
+        return 200;
+    }
+    return 500 + 100 * (std::abs(dy - sy)/4 + std::abs(dx - sx)/4);
+}
+
+
+delay_t NgUltraImpl::predictDelay(BelId src_bel, IdString src_pin, BelId dst_bel, IdString dst_pin) const
+{
+    Loc src_loc = ctx->getBelLocation(src_bel), dst_loc = ctx->getBelLocation(dst_bel);
+    if (src_loc.x==dst_loc.x && src_loc.y==dst_loc.y) {
+        // Same sub tile
+        return 50;
+    } else if (((src_loc.x & 0xfffc) == (dst_loc.x & 0xfffc)) && ((src_loc.y & 0xfffc) == (dst_loc.y & 0xfffc))) {
+        // Same "TILE"
+        return 200;
+    }
+    return 500 + 100 * (std::abs(dst_loc.y - src_loc.y)/4 + std::abs(dst_loc.x - src_loc.x)/4);
+}
+
+void NgUltraImpl::fixup_crossbars()
+{
+
+    auto is_crossbar_pip = [&] (PipId pip) {
+        const auto &pd = chip_pip_info(ctx->chip_info, pip);
```

Comment:
maybe we should have a convenience function somwhere that does the `const auto &pd = chip_pip_info(ctx->chip_info, pip);`, `const auto &extra_data = *reinterpret_cast<const NGUltraPipExtraDataPOD *>(pd.extra_data.get());` as this seems to come up in lots of places